doc/pins/p4rt_app_hld.md

```diff
@@ -0,0 +1,169 @@
+* Allow P4RT table entries in the APPL_DB to be read back.
+* Support Packet I/O for punt rules defined by P4RT (TODO: reference PacketIO doc).
+
+
+## P4RT Application High-Level Design
+
+The P4RT application is responsible for implementing the [P4Runtime Specification](https://p4lang.github.io/p4runtime/spec/v1.3.0/P4Runtime-Spec.html) as a gRPC service. The application will run in its own Docker container. A controller (or multiple controllers) can connect as clients to the gRPC service, and write P4 table entries that correspond to SAI tables (specified by a P4 program). These entries are intended for the ASIC database, but following the SONiC architecture, the P4RT application will translate/format the gRPC request and write its output into the APPL_DB. From there, the P4Orch (TODO: reference OA HLD) handles the request.
+
+![drawing](images/p4rt_app.png)
+
+
+### Arbitration & Controller Roles
+
+As part of the P4RT specification, [multiple controllers can connect to the gRPC service](https://p4lang.github.io/p4runtime/spec/v1.3.0/P4Runtime-Spec.html#sec-client-arbitration-and-controller-replication). Each controller can have a unique role which limits its access to specific tables. A default role can also be used which gives full pipeline access. Multiple connections for the same role are allowed. In this case, the P4RT application will arbitrate between all the connections using an election ID provided by each client. A single primary connection is selected for each role and that client is allowed to update tables; all other connections become backups which have read-only access to the pipeline.
+
+
+### P4 programs & P4Info
+
+![drawing](images/sai_p4.png)
+
+The PINS effort will model the SAI pipeline as a P4 program which will be compiled using a [P4 compiler](https://github.com/p4lang/p4c). The compiler generates a P4Info file which describes all parts of the SAI pipeline controllable from the P4RT application (e.g. tables, counters, or groups).
+
+When a SDN controller first connects to the P4RT application, it will push the P4Info. The application verifies the config provided by the P4Info and then realizes it on the data plane. Any failure from P4RT application or a lower layer to realize a part of the config means it will be rejected. Verification will check:
+
+* Required components are included in the P4Info. For example controller metadata about PacketIO matches expectations.
+* Field Types around Fixed SAI components are supported (e.g. ports or nexthop IDs are passed as strings, etc.).
+* Configurable SAI components (e.g. ACL tables) will write their definitions into the APPL_DB, and the OA layer will tell P4RT application if they are accepted.
+
+
+### ACL Table Definitions
+
+The configurable parts of the SAI pipeline are switch dependent, and must respect hardware limitations. These limitations can be difficult to verify at the P4 program level, but are much simpler to verify on actual hardware. For example:
```

Comment:
Could you give example of "The configurable parts of the SAI pipeline are switch dependent"?